### ISSUE TITLE
improve use of siconos as a library. Siconos error handling.

### ISSUE THREAD

**Issue by xhub** [siconos/siconos#97]
It's great that we have a concrete example where siconos could be use as a library by other softs.
However, I think that we could improve on the following topics:
- we use exit() a lot of time when we encounter an error that we cannot deal with.
- we use printf/fprintf a lot to report errors. However, it is sometimes not ideal, and when using multiples threads, the output is all mixed-up. We should be able to save those messages to a file, or to report it to a logging mechanism that the user can set.
- we still have some global variables. My understanding is that we could convert a lot of those to thread_local variables (see tlsdef.h). This means that they are global, but only to a given thread (and not all of them).

There exists a standard mechanism (ANSI IIRC) for handling errors: setjmp/longtmp, which is known under the not-so-sexy name of non-local goto. See
- man setjmp 
- http://www.di.unipi.it/~nids/docs/longjump_try_trow_catch.html
- https://en.wikipedia.org/wiki/Setjmp.h
- https://en.wikibooks.org/wiki/C_Programming/Coroutines#setjmp 

Regarding the fatal error handling, I propose the following:
- make sure that in the solvers (the function called in the driver), there is no exit() call, but that they always return (and the return code indicate the error).
- Provide a standard error codes with a function to explain it (e.g. return a char* to a description of the error).
- use setjmp/longjmp to provide a mechanism to deal with fatal errors in driver functions: if there is a fatal error during the execution, the program is going to jump back to the driver function and return properly.
- ban the use of exit()

We would introduce a fatal_error() function that would do the following:
- call the user-defined error handler if one has been set
- call longjmp if the jmp_buf has been initialized
- call abort(): this function sends a SIGABRT signal first, that can be intercepted if the user sets a signal handler.

Basically all the function in numerics/src/tools (+ the ones that computes error / projection) could call the fatal_error function. All the solver function must return. We would then advertise that only the driver functions have out of the box error handling, and that a custom error handling is available.

What do we gain from this: better use of siconos as a library and of the python bindings (ipython would no longer exit when we have a fatal error, a MATLAB interface would also benefit from that), and a better way to log messages.

**Comment by vacary** [comment]
Fully agrée with all your propositions. Great. 

Is it possible to implement the solution for one critical exit in the code and then we duplicate

For the function that interprets the error code it is generally very convenient. The mechanism in the swig implementation  is very useful for instance

**Comment by radarsat1** [comment]
I think this is a good idea although it represents a lot of work to do, but probably worth it.

The only point on which I would exercise caution is specifying a global error handler.  We have no "Siconos context" object in which to register a handler, so it would have to be global.  (Unless it is something we could pass into Simulation and down into numerics.. but that sounds complicated.)  In general I prefer returning error codes to calling handlers when it comes to fatal errors.  Not so much for C++ programming practices but because it's easier to use when binding the library for other languages.  Indeed we'd have to make sure anything we do here is nicely Python-compatible.

If we do register a callback, let's make sure it has a field or parameter for a user struct or user context pointer so that it can be easily associated with data.

**Comment by radarsat1** [comment]
In terms of error codes, it would be good to have some standard set of return values, like 0 for no error and otherwise something else.  In fact new C++14 standards has std::optional, although it's present in boost as well.  Not sure we should use that however, but something along those lines can be very useful, like returning an error code and providing a way to get more information on the error.

As for printing messages, I'm for adding a (sane and fast) logging mechanism.  Not sure of the best way forward on that.  We talked about [boost::log](https://theboostcpplibraries.com/boost.log) previously but I haven't really evaluated it yet or compared to other solutions.  Again, I see context structures being a potential issue here.

i.e. Should logging be "global" by nature? or record to a provided memory structure of some sort..  Perhaps we will need some kind of context structure to associate with the Simulation to hold things like logging and error information.

**Comment by xhub** [comment]
> Is it possible to implement the solution for one critical exit in the code and then we duplicate

I'll implement along those lines in ncp_driver to show a first draft,
and I may modify one function in NumericsMatrix.c

> For the function that interprets the error code it is generally very convenient. The mechanism in the swig implementation  is very useful for instance

I'm not sure I know what you're refering to.

**Comment by xhub** [comment]
> I think this is a good idea although it represents a lot of work to do, but probably worth it.

I don't think it is so much work: sed will do most of the work. I plan to change (f)printf() to a logging function, and exit() to a siconos_fatal_error()

> The only point on which I would exercise caution is specifying a global error handler.  We have no "Siconos context" object in which to register a handler, so it would have to be global.  (Unless it is something we could pass into Simulation and down into numerics.. but that sounds complicated.)  In general I prefer returning error codes to calling handlers when it comes to fatal errors.  Not so much for C++ programming practices but because it's easier to use when binding the library for other languages.  Indeed we'd have to make sure anything we do here is nicely Python-compatible.

Just to be precise, I will use thread local storage for those variables.
I'm not sure what you mean by "Siconos context object". A struct that we be passed everywhere?
I fully agree that returning an error code is the way to code. However, it would be a massive work to modify all the code to ensure that we can easily return to the code calling 

> If we do register a callback, let's make sure it has a field or parameter for a user struct or user context pointer so that it can be easily associated with data.

Yes. I don't have a specific idea for it. I wanted to ask if Gazebo had something along those lines for handling errors in external libraries.

In python/swig, there is a handy %exception statement that can be used to wrap all the function calls. Then we only need to set the jmp_buf from python and then it does the trick.  In MATLAB, I plan on doing the same.

**Comment by radarsat1** [comment]
>  I'm not sure what you mean by "Siconos context object". A struct that we be passed everywhere?

To be clear, I don't want this either!  But since we were discussing getting rid of globals, it seems to me the logging function would need some "object" that stores the messages somewhere, same for reporting extra error information. Maybe it's not problematic.  In any case there are no programs that I'm aware of that would have two Simulations for which you'd want to separate the error messages, but it's not unthinkable.

As for Gazebo, it doesn't expect any kind of catastrophic failures in external libraries.  It does use a very C++-y logging system, e.g.,

~~~
  // If only the parent exists, then create a joint between the parent
  // and the world.
  else if (bulletParentLink)
  {
    this->bulletSlider = new btSliderConstraint(
        *bulletParentLink->GetBulletLink(), frameParent, true);
  }
  // Throw an error if no links are given.
  else
  {
    gzerr << "joint without links\n";
    return;
  }
~~~

It is something along the lines of this "gzerr" object that I was meaning as a context for where to store things.  In Gazebo, it is indeed a [static global object named "err"](https://osrf-distributions.s3.amazonaws.com/gazebo/api/dev/classgazebo_1_1common_1_1Console.html#details) of type `Logger`.

**Comment by radarsat1** [comment]
I'll add that `gzerr` is defined like this:
~~~
/// \brief Output an error message
    #define gzerr (gazebo::common::Console::err(__FILE__, __LINE__))
~~~
so that the source location of the error message can be logged.

**Comment by xhub** [comment]
Ok I realized I forgot to mention that I was thinking of numerics when I created this issue, and not at all of the C++ part of siconos. My bad. I don't think it makes sense to use setjmp/longjmp in C++. Also for the logging, I think there are other options in C++.

> In terms of error codes, it would be good to have some standard set of return values, like 0 for no error and otherwise something else.  In fact new C++14 standards has std::optional, although it's present in boost as well.  Not sure we should use that however, but something along those lines can be very useful, like returning an error code and providing a way to get more information on the error.

Unfortunately, I don't think using c++14 is an option right now. Also I'm thinking more of numerics and I'd like to have a solution it in C.

> As for printing messages, I'm for adding a (sane and fast) logging mechanism.  Not sure of the best way forward on that.  We talked about [boost::log](https://theboostcpplibraries.com/boost.log) previously but I haven't really evaluated it yet or compared to other solutions.  Again, I see context structures being a potential issue here.
> 
> i.e. Should logging be "global" by nature? or record to a provided memory structure of some sort..  Perhaps we will need some kind of context structure to associate with the Simulation to hold things like logging and error information.

I'm going to focus on numerics for the forseable future. If someone else wants to handle this for the C++ part, feel free to do so.

For numerics, by default, we would output on stdout or stderr.  If the user register a function that takes a char* and an int (for the severity), it is called.  Otherwise we could log to a file. I would avoid logging with multi-threading. Let's see if it is possible to do that for a start.

**Comment by radarsat1** [comment]
Ah that makes sense. I was thinking about Kernel more. You're right it's better to have something that works well for logging in C and maybe we can leverage it nicely in C++. I have some experience wrapping C callbacks with C++ objects, all it takes is an extra void* context pointer so that you can pass a function that calls an object.

**Comment by xhub** [comment]
> Ah that makes sense. I was thinking about Kernel more. You're right it's better to have something that works well for logging in C and maybe we can leverage it nicely in C++. I have some experience wrapping C callbacks with C++ objects, all it takes is an extra void* context pointer so that you can pass a function that calls an object. 

Hm, I'm not sure I can pretend to do something nice in the near future :)

**Comment by radarsat1** [comment]
Yup, I meant, feel free to work on the C part. :)  We'll build something sensible on top of it!

**Comment by radarsat1** [comment]
Er... something sensible for C++.. not to imply what you will do in C is not sensible.. lol.. sorry ;)

**Comment by radarsat1** [comment]
Recently I ran a big simulation for testing, and it brings up an issue that I think is related to this one: I cannot interrupt the solver using Ctrl-C!  It would be nice if it were possible to interrupt the solver loop and have it exit properly to the Python interpreter without segfaulting. I was confused by this because I don't see any signal() capture in numerics or in SWIG.  Now I thought about it more, and I think maybe Ctrl-C gets blocked by the Python interpreter when it calls out to extension code.  Some searching reveals that this is not an unknown issue for long-running C extensions, and has been handled in various ways by different solutions.  I propose that we can perhaps borrow some intelligent signal handling from Cython:

http://cysignals.readthedocs.io/en/latest/sigerror.html

http://cysignals.readthedocs.io/en/latest/interrupt.html

**Comment by xhub** [comment]
I believe we should add a callback for signal checking somewhere in the code. For python, PyErr_CheckSignals https://docs.python.org/3.5/c-api/exceptions.html#c.PyErr_CheckSignals should be periodically checked. For other programming languages, we should see. This way the can bail out cleanly.

What is the "solver loop"?

**Comment by radarsat1** [comment]
That's not a bad idea.  We could have some kind "SiconosErrorHandler" object?  Perhaps a simple callback would suffice.  By "solver loop" I mean the GS loop, i.e, the time between when OSNS::compute() is called and it returns.

**Comment by xhub** [comment]
I was thinking of a callback. In numerics, we could periodically check whether a signal was given. That could be the Newton loop, the GS loop or any other loop.

**Comment by radarsat1** [comment]
Yeah of course it would have to be a callback+context pointer for the C case, makes sense.  Do you think it would fit nicely in SolverOptions, or should it go elsewhere? It would be nice to avoid adding a global imho.

Ah, indeed now that I look, there is already a Callback struct, I guess we could just add it there.

**Comment by xhub** [comment]
I would set this in a TLS pointer to function. I am fine with this object being unique to the thread, since we may not switch between python/matlab or standalone. It would be nice to have that set whenever the python module is loaded.

I would envision something like

if (signal_handler && iter % 10 == 0) { int has_sig = signal_handler(); if (has_sig) goto _exit_nicely; }

**Comment by xhub** [comment]
Just by reading it, it looks like we had an idea of implementing a signal handler in some solver loop. This is still worth looking into it. I think it belongs to an other issue maybe

**Comment by vacary** [comment]
Transferred in wiki

https://github.com/siconos/siconos/wiki/Wish-list-and-enhancement
